docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,410 @@
+
+The contract consists of three resources:
+
+- An IPAddressClaim, which is used to allocate addresses
+- An IPAddress that IPAM providers use to fulfil IPAddressClaims
+- IP Pools, which hold provider specific configuration on how addresses should be allocated, e.g. a specific subnet to use
+
+Both **IPAddressClaims** and **IPAddresses** should be part of Cluster API, while **IPPools** are defined by the different IPAM providers.
+
+An **IPAddressClaim** is used by infrastructure providers to request an IP Address. The claim contains a reference to an IP Pool. Because the pool is provider specific, the IPAM controllers can decide whether to handle a claim by inspecting the group and kind of the pool reference.
+
+If a IPAM controller detects a Claim that references a Pool it controls, it allocates an IP address from that pool and creates an **IPAddress** to fulfil the claim. It also updates the status of the IPAddressClaim with a reference to the created Address.
+
+```go
+// IPAddressClaimSpec describes the desired state of an IPAddressClaim
+type IPAddressClaimSpec struct {
+   // Pool is a reference to the pool from which an IP address should be allocated.
+   Pool LocalObjectReference `json:"pool,omitempty"`
+}
+
+// IPAddressClaimStatus contains the status of an IPAddressClaim
+type IPAddressClaimStatus struct {
+   // Address is a reference to the address that was allocated for this claim.
+   Address LocalObjectReference `json:"address,omitempty"`
+}
+ 
+// IPAddressClaim can be used to allocate IPAddresses from an IP Pool.
+type IPAddressClaim struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+ 
+   Spec   IPAddressClaimSpec   `json:"spec,omitempty"`
+   Status IPAddressClaimStatus `json:"status,omitempty"`
+}
+```
+
+```go
+// IPAddressSpec describes an IPAddress
+type IPAddressSpec struct {
+   // Claim is a reference to the claim this IPAddress was created for.
+   Claim LocalObjectReference `json:"claim,omitempty"`
+ 
+   // Pool is a reference to the pool that this IPAddress was created from.
+   Pool LocalObjectReference `json:"pool,omitempty"`
+ 
+   // Address is the IP address.
+   Address string `json:"address"`
+ 
+   // Prefix is the prefix of the address.
+   Prefix int `json:"prefix,omitempty"`
+ 
+   // Gateway is the network gateway of network the address is from.
+   Gateway string `json:"gateway,omitempty"`
+}
+ 
+// IPAddress is a representation of an IP Address that was allocated from an IP Pool.
+type IPAddress struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+ 
+   Spec IPAddressSpec `json:"spec,omitempty"`
```

Comment:
do we foresee the need for status here? e.g could an IP be release and reused by a different claim?

Reply:
I don't really see a need to reuse an IPAddress object. It's not like Volumes where there is persistent data associated with the object, which is lost when it's deleted and replaced with an spec-wise identical one. The IP address continues to exist, it's just not allocated anymore.
With the current scope of the proposal, which is about integration for machines, which are dynamically created and deleted, the only use case I can think of would be pre-allocating special addresses that can then be specifically selected, but that's something that's pretty provider specific and contradicts the "request -> fulfilment" approach. And it can also be implemented by providers using annotations.

Another use case might be fixed addresses for e.g. the api server. But those static addresses can also be set on cluster creation, and don't really require the dynamic allocation proposed here.